### ISSUE TITLE
C3002 - False positive?  Possibly related to lambda inside lamba function

### ISSUE THREAD

**Issue by mdeweerd** [PyCQA/pylint#7308]
### Bug description

```raw
example.py:12:42: C3002: Lambda expression called directly. Execute the expression inline instead. (unnecessary-direct-lambda-call)
```

The line it is reported for is:
```
    return sorted(items, key=lambda item: (lambda a: (a is None, a))(item[sort_key]))
```

I can't see how to factor out a lamba function without repeating `item[sort_key]` .
Note: in the original code `item[sort_key]` is slightly more complex.

Full example code:
```python
#!/usr/bin/env python3
"""Example for potentially false positive for pylint C3002 notification"""
from __future__ import annotations

import os

from var_dump import var_dump  # pip package


def get_sorted_list_of_dict(
    items: list[dict[str, None | str | int]], sort_key: str
) -> list[dict[str, None | str | int]]:
    """Sort list according to sort_key in dict"""
    return sorted(items, key=lambda item: (lambda a: (a is None, a))(item[sort_key]))


test_list = [
    {"k": 3, "v": "A"},
    {"k": None, "v": "B"},
    {"k": 1, "v": "C"},
    {"k": 5, "v": "D"},
    {"k": 0, "v": "E"},
    {"k": 4, "v": None},
]

print("Original list" + os.linesep)
var_dump(test_list)
print("Sorted by 'k'" + os.linesep)
var_dump(get_sorted_list_of_dict(test_list, "k"))
print("Sorted by 'v'" + os.linesep)
var_dump(get_sorted_list_of_dict(test_list, "v"))
```

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:14:42: C3002: Lambda expression called directly. Execute the expression inline instead. (unnecessary-direct-lambda-call)

------------------------------------------------------------------
Your code has been rated at 9.17/10 (previous run: 9.17/10, +0.00)
```

### Expected behavior

No C3002 warning 

### Pylint version

```shell
pylint 2.14.5
astroid 2.11.6
Python 3.8.12 (default, Nov 23 2021, 20:18:25)
[GCC 11.2.0]
```

### OS / Environment

Cygwin

### Additional dependencies

agate==1.6.3
agate-dbf==0.2.2
agate-excel==0.2.4
agate-sql==0.5.7
aiodns==3.0.0
aiohttp==3.8.1
aiosignal==1.2.0
aiosqlite==0.17.0
appdirs @ file:///pub/devel/python/python-appdirs/python-appdirs-1.4.4-2.noarch/build/dist/appdirs-1.4.4-py2.py3-none-any.whl
astroid==2.11.6
async-timeout==4.0.2
asynctest==0.13.0
attr==0.3.1
attrs==21.4.0
autoflake8==0.3.0
Babel==2.9.0
bandit==1.7.2
beautifulsoup4 @ file:///pub/devel/python/python-bs4/python-bs4-4.10.0-1.noarch/build/dist/beautifulsoup4-4.10.0-py3-none-any.whl
black==22.1.0
certifi==2020.12.5
cffi @ file:///pub/devel/python/python-cffi/python-cffi-1.15.0-1.x86_64/build/dist/cffi-1.15.0-cp38-cp38-cygwin_3_3_3_x86_64.whl
cfgv==3.3.1
chardet==3.0.4
charset-normalizer==2.0.12
click==8.0.1
codespell==2.1.0
coloredlogs==15.0.1
coverage==6.3.1
coveralls==3.3.1
crccheck==1.1
csvkit==1.0.6
dbfread==2.0.7
dill==0.3.5.1
dispy==4.15.1
distlib==0.3.1
distro==1.6.0
docopt==0.6.2
dparse==0.5.1
et-xmlfile==1.1.0
filelock==3.6.0
flake8==4.0.1
flake8-2020==1.6.1
flake8-bugbear==22.1.11
flake8-comprehensions==3.8.0
frozenlist==1.3.0
future==0.18.2
git-filter-repo==2.33.0
gitdb==4.0.9
GitPython==3.1.26
greenlet==1.1.0
html5lib @ file:///pub/devel/python/python-html5lib/python-html5lib-1.1-2.noarch/build/dist/html5lib-1.1-py2.py3-none-any.whl
humanfriendly==10.0
identify==2.4.9
idna==2.10
importlib-metadata==4.11.0
importlib-resources==5.7.1
iniconfig==1.1.1
iso8601==0.1.13
isodate==0.6.0
isort==5.10.1
jsonschema==4.5.1
lazy-object-proxy==1.7.1
leather==0.3.3
linkify-it-py==1.0.3
lockfile @ file:///pub/devel/python/python-lockfile/python-lockfile-0.12.2-4.noarch/build/dist/lockfile-0.12.2-py2.py3-none-any.whl
lxml==4.6.3
markdown-it-py==2.0.1
mccabe==0.6.1
mdformat==0.7.13
mdformat-gfm==0.3.5
mdformat-toc==0.3.0
mdformat_frontmatter==0.4.1
mdformat_tables==0.4.1
mdit-py-plugins==0.3.0
mdurl==0.1.0
mistune==0.8.4
mock==4.0.3
multidict==6.0.2
mypy-extensions==0.4.3
nodeenv==1.6.0
numpy==1.22.2
olefile==0.46
openpyxl==3.0.7
packaging==20.9
parsedatetime==2.4
pathspec==0.9.0
pbr==5.8.1
PcbDraw==0.6.1
pip-autoremove==0.10.0
platformdirs==2.5.0
pluggy==0.13.1
ply @ file:///pub/devel/python/python-ply/python-ply-3.11-3.noarch/build/dist/ply-3.11-py2.py3-none-any.whl
pre-commit==2.19.0
py==1.10.0
py-enigma==0.1
pybars3==0.9.7
pycares==4.1.2
pycodestyle==2.8.0
pycos==4.12.1
pycparser @ file:///pub/devel/python/python-pycparser/python-pycparser-2.20-2.noarch/build/dist/pycparser-2.20-py2.py3-none-any.whl
pycrypto @ file:///pub/devel/python/python-crypto/python-crypto-2.6.1-4.x86_64/build/dist/pycrypto-2.6.1-cp38-cp38-cygwin_3_3_3_x86_64.whl
pycryptodome==3.14.1
pyflakes==2.4.0
pylint==2.14.5
PyMeta3==0.5.1
pyOpenSSL @ file:///pub/devel/python/python-openssl/python-openssl-21.0.0-1.noarch/build/dist/pyOpenSSL-21.0.0-py2.py3-none-any.whl
pyparsing==2.4.7
pyrsistent==0.18.1
pyserial==3.5
pyserial-asyncio==0.6
pytest==7.0.1
pytest-aiohttp==1.0.4
pytest-asyncio==0.18.1
pytest-cov==3.0.0
pytest-timeout==2.1.0
python-slugify==6.1.1
pytimeparse==1.1.8
pytz==2021.1
pyupgrade==2.31.0
PyYAML==6.0
regex==2021.4.4
requests==2.23.0
requests-mock==1.9.3
ruamel.yaml==0.17.21
ruamel.yaml.clib==0.2.6
ruyaml==0.91.0
safety==1.10.3
serial==0.0.97
six @ file:///pub/devel/python/python-six/python-six-1.16.0-1.noarch/build/dist/six-1.16.0-py2.py3-none-any.whl
smmap==5.0.0
solid==0.2.0
SQLAlchemy==1.4.22
stevedore==3.5.0
svgutils==0.3.4
swagger-client==1.0.0
text-unidecode==1.3
tokenize-rt==4.2.1
toml==0.10.2
tomli==2.0.1
tomlkit==0.11.0
tox==3.24.5
typed-ast==1.5.2
types-PyYAML==6.0.4
types-requests==2.27.9
types-setuptools==57.4.9
types-six==1.16.10
types-urllib3==1.26.9
typing_extensions==4.0.1
uc-micro-py==1.0.1
urllib3==1.25.11
var-dump==1.2
virtualenv @ file:///pub/devel/python/python-virtualenv/python-virtualenv-20.13.0-1.noarch/build/dist/virtualenv-20.13.0-py2.py3-none-any.whl
voluptuous==0.12.2
Wand==0.6.6
webencodings @ file:///pub/devel/python/python-webencodings/python-webencodings-0.5.1-4.noarch/build/dist/webencodings-0.5.1-py2.py3-none-any.whl
wrapt==1.14.1
xlrd==2.0.1
xlsx2csv==0.7.8
yamlfix==0.8.0
yamllint==1.26.3
yarl==1.7.2
zigpy==0.43.0
zigpy-znp==0.7.0
zipp==3.7.0

**Comment by DanielNoord** [comment]
Just to be clear, you're using:
```python
lambda item: (lambda a: (a is None, a))(item[sort_key]))
```

To avoid:
```python
lambda item: (item[sort_key] is None, item[sort_key])
```

RIght?

**Comment by mdeweerd** [comment]
Basically, yes.  In my case 'item[key]' is somewhat more complex, so this is to avoid repeating code (DNR principle) and especially avoid repeating more complex expressions.

**Comment by DanielNoord** [comment]
Not sure if we can "fix" this as this really comes down to style. Personally I would suggest to store the complex variable in a temp variable and just re-use this. Imo, nesting lambda's like this doesn't really contribute to readability but I fully understand this is a personal preference.

That said, I don't know if we could come up with a good heuristic for when lambda's use too intricate "item[key]'s" and when we shouldn't emit this warning. I think perhaps the best solution is thus to disable it for this specific line with a `pylint: disable` or `disable-next` comment. Unless you have a good suggestion for a heuristic of course!

**Comment by DanielNoord** [comment]
In that case:
```python
lambda item: (lambda a: (a is None, a))(item[sort_key]))
```
Would no longer flag, which is arguably exactly one of the type of statements that this checker is trying to avoid.

**Comment by mdeweerd** [comment]
The pylint notification is "unnecessary-direct-lambda-call" .  It can be argued that no lamba call is ever necessary - they did not exist in most languages.

"unnecessary": it can be argued that it is necessary because it avoids code repetition which is IMHO of more importance than adding a lambda function.

Is it a direct call ?  No really IMHO because the parameter is complex.

```python
lambda item: (lambda a: (a is None, a))(item))
```
would be overkill as it can be successively optimised to:

```python
lambda item: (item is None, item)
(item is None, item)
```

Supposing that the checker is trying to avoid
```python
( lamba a: any_function_of(a) ) (parameter)
```

Then I am ok with that if 'parameter' is just a value or a simple variable.
But as soon as paramter is more complex than that, then the "Do Not Repeat" principel comes into play.  And that avoids bugs introduces by code modifications that are made in one place but not in the other.

The lamba expression with an argument arguably better expresses what you are doing with the argument than when you are repeating the same complex argument multiple times in an expression.  Which is one reason to have functions.

Arguably one could write

```python
def get_sorted_list_of_dict(
    items: list[dict[str, None | str | int]], sort_key: str
) -> list[dict[str, None | str | int]]:
    """Sort list according to sort_key in dict"""
    # return sorted(items, key=lambda item: (lambda a: (a is None, a))(item[sort_key]))
    def falsify_none(arg):
        return (arg is None, arg)
    return sorted(items, key=lambda item: falsify_none(item[sort_key]))
```

Or, even remove all lambda functions because having lambda functions as direct function parameters is argually bad, but possibly pythonic:

```python
def get_sorted_list_of_dict(
    items: list[dict[str, None | str | int]], sort_key: str
) -> list[dict[str, None | str | int]]:
    """Sort list according to sort_key in dict"""
    # return sorted(items, key=lambda item: (lambda a: (a is None, a))(item[sort_key]))
    def my_compare_function(item):
        def falsify_none(arg):
            return (arg is None, arg)
        return falsify_none(item[sort_key])
    return sorted(items, key=my_compare_function)
```

So basically I am in favor that this rule only applies when the parameter is not complex as in that case there is no argument that the lambda function is unnecessary.

It can be argued that [symilar](https://pylint.pycqa.org/en/latest/symilar.html) would have to flag `lambda item: (item[sort_key] is None, item[sort_key])` because `item[sort_key]` is repeated and that repetition can be avoided.

**Comment by DanielNoord** [comment]
> The pylint notification is "unnecessary-direct-lambda-call" . It can be argued that no lamba call is ever necessary - they did not exist in most languages.

This is not about the existence of lambda's but rather about calling a lambda immediately after defining it.
For example, one of our test cases is:

`y = (lambda x: x**2 + 2*x + 1)(a)` which should just be `y = a**2 + 2*a + 1`. There is no need to define a lambda if you're just going to call it immediately. If anything it will be less performant code.

> "unnecessary": it can be argued that it is necessary because it avoids code repetition which is IMHO of more importance than adding a lambda function.

Honestly this is debatable. For very complex repeated statements this might be true, but in other cases a little bit of repetition to save the performance impact of assigning the lambda might be desired. There is no single truth here I think.

> Is it a direct call ? No really IMHO because the parameter is complex.

I mean, this is not really debatable. The call is direct, there are parentheses just after the lambda definition.

> Then I am ok with that if 'parameter' is just a value or a simple variable. But as soon as paramter is more complex than that, then the "Do Not Repeat" principel comes into play. And that avoids bugs introduces by code modifications that are made in one place but not in the other.

Again, I'm not disagreeing here but I'm not sure there is simple heuristic to determine "complexity". Is that the amount of characters? The amount of nested lookups? Is that the repetition of hard to distinguish letters such as I and l?
If you're really worried about messing up repeated references to a complex paramater why not define a temporary variable? That's better for code readability anyway.

_I have no idea but I could even imagine that defining a variable is faster than defining a lambda. For anecdotal evidence see:_
```console
❯ python -m timeit 'x = 1'
50000000 loops, best of 5: 9.84 nsec per loop
❯ python -m timeit 'y = lambda x: x'
5000000 loops, best of 5: 64.1 nsec per loop
```

> The lamba expression with an argument arguably better expresses what you are doing with the argument than when you are repeating the same complex argument multiple times in an expression. Which is one reason to have functions.

If defining a function improves the readability and robustness of the code that might indeed be better.

> So basically I am in favor that this rule only applies when the parameter is not complex as in that case there is no argument that the lambda function is unnecessary.

See above, I don't think there is a good way to determine complexity and I think we should be a little bit more strict here. `unnecessary-...` messages will always create a little bit of debate but the example you gave (I know that it is not your actual production code) is exactly what this checker wants to prevent.
You're right that there are cases where it might indeed be better to do a direct lambda call for those who are used to such syntax. However, in general I think the checker is still valid. In those edge cases where it isn't I think it's better to ask the user to add an `# pylint: disable` than to try and come up with an arbitrary way to measure "complexity" (which will likely also receive criticism from another group of users).

> It can be argued that [symilar](https://pylint.pycqa.org/en/latest/symilar.html) would have to flag `lambda item: (item[sort_key] is None, item[sort_key])` because `item[sort_key]` is repeated and that repetition can be avoided.

If that were suggested I'd vote against it.

**Comment by mdeweerd** [comment]
> `y = (lambda x: x**2 + 2*x + 1)(a)` which should just be `y = a**2 + 2*a + 1`
Sure we agree on that, possibly not entirely for the same reasons - in my case it's just because `a` is "simple".

# "complexity".

> Is that the amount of characters? The amount of nested lookups? Is that the repetition of hard to distinguish letters such as I and l?

Complexity is not measured in terms of a number of characters.

In this context, anything else than a variable, a constant of a numerical value or a numeric value not in an expression such as multiplication, division, list, array indexing, function call.

# Performance wise:

A fairer comparison, pointing even more to the computational inefficiency of lambda functions in python,  would be:
```shell
python3 -OO -m timeit "a = 1; x = 1"
python3 -OO -m timeit "a = 1; x = (lambda y: y)(a);"
```

The difference in speed is a Python's compiler problem because it could eventually optimize this out once it's mature enough, and Python is just not in itself efficient as it is an interpreted language.  To me, the focus is on coding effort.

Anyway, you're not convinced on this and I am not going to spend more effort arguing this.

Thanks for pylint.

**Comment by DanielNoord** [comment]
> Anyway, you're not convinced on this and I am not going to spend more effort arguing this.
> 
> Thanks for pylint.

Thanks you for participating in the discussion and not just getting mad. We have seen many examples of other behaviour sadly 😅
